core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1496,37 +1513,36 @@ protected void operationProcessed() {
             mapperService.indexAnalyzer(), similarityService.similarity(mapperService), codecService, shardEventListener, translogRecoveryPerformer, indexCache.query(), cachingPolicy, translogConfig, inactiveTime);
     }
 
-    private static class IndexShardOperationCounter extends AbstractRefCounted {
-        final private ESLogger logger;
-        private final ShardId shardId;
-
-        public IndexShardOperationCounter(ESLogger logger, ShardId shardId) {
-            super("index-shard-operations-counter");
-            this.logger = logger;
-            this.shardId = shardId;
-        }
-
-        @Override
-        protected void closeInternal() {
-            logger.debug("operations counter reached 0, will not accept any further writes");
+    public Releasable newOperationLock() {
+        verifyNotClosed();
+        if (state == IndexShardState.RELOCATED && shardRouting.primary()) {
+            throw new IndexShardRelocatedException(shardId);
         }
-
-        @Override
-        protected void alreadyClosed() {
-            throw new IndexShardClosedException(shardId, "could not increment operation counter. shard is closed.");
+        // <-- position X
+        Releasable releasable;
+        try {
+            releasable = suspendableRefContainer.tryAcquire();
+        } catch (InterruptedException e) {
+            logger.warn("operation lock was interrupted");
+            Thread.currentThread().interrupt(); // Mark thread as interrupted again
```

Comment:
we throw the exception and thus take care of the interrupt. We don't need to set it...